include/revng/Model/Types.h

```diff
@@ -0,0 +1,903 @@
+
+  StructField(StructField &&) = default;
+  StructField &operator=(StructField &&) = default;
+
+  bool operator==(const StructField &) const = default;
+};
+INTROSPECTION_NS(flatc, StructField, FieldType, Name, Offset);
+
+/// \brief KeyedObjectTraits for flatc::StructType based on its byte-offset in
+/// struct
+template<>
+struct KeyedObjectTraits<flatc::StructField> {
+
+  static std::size_t key(const flatc::StructField &Val) { return Val.Offset; }
+
+  static flatc::StructField fromKey(const std::size_t &Offset) {
+    return flatc::StructField(Offset);
+  }
+};
+
+/// \brief A struct type in flatc. Structs are actually typedefs of unnamed
+/// structs in C.
+struct flatc::StructType : public flatc::Type {
+  static constexpr const char *Tag = "!flatc::StructType";
+
+  // StructType is kinda weird, because we have a set of conflicting
+  // requirements:
+  // - first of all, AggregateFields contains a TupleTreeReference, so we
+  // cannot
+  //   hold fields in a std::vector
+  //   - this is due to the fact that TupleTree still does not handle
+  //   automatic initialization of TupleTreeReferences that are inside a
+  //   std::vector
+  //   - so we need to hold the fields in a supported container. likely a
+  //   SortedVector. but we have to decide the key that will be used to order
+  //   them
+  // - we want to be able to index fields by name
+  //   - this is a hint that the Name should be the key for fields
+  //   - this is also consistent with the fact that in C it's not possible to
+  //   have 2 different fields of a struct with the same name
+  //   - but accessing fields by name is rarely necessary in practice
+  //   - also, renaming a field changes its order
+  // - most times we want to iterate on fields we actually want to do it in
+  //   positional order, meaning that fields that come first in the actual
+  //   layout of the struct would need to be iterated over sooner
+  //   - this suggests that we want the key for the fields to be an integer
+  //   representing their position in the struct
+  //   - but if we remove a field from the struct after its creation, we end
+  //   up with a missing index, which is a thing that we'd like to avoid
+  //   - this kind of iteration is important for actually computing pointer
+  //   arithmetic and member access expressions when generating C code, and
+  //   it's likely the most common operation
+
+  SortedVector<StructField> Fields;
+  std::size_t Size;
+
+  StructType(llvm::StringRef NameRef) :
+    Type(NameRef, TypeKind::Struct), Fields() {}
+
+  StructType() : StructType("") {}
+  ~StructType() = default;
+
+  StructType(const StructType &) = default;
+  StructType &operator=(const StructType &) = default;
+
+  StructType(StructType &&) = default;
+  StructType &operator=(StructType &&) = default;
+
+  static bool classof(const Type *T) {
+    return T->getKind() == TypeKind::Struct;
+  }
+
+  RecursiveCoroutine<std::size_t> size() const;
```

Comment:
Can we push all the `RecursiveCoroutine` into the `.cpp` file and drop `#include "revng/ADT/RecursiveCoroutine.h"`?